src/algorithm/REDUCE_SUM-Cuda.cpp

```diff
@@ -0,0 +1,257 @@
+                                           m_sum_init,
+                                           stream));
+
+    // Allocate temporary storage
+    unsigned char* temp_storage;
+    allocCudaDeviceData(temp_storage, temp_storage_bytes);
+    d_temp_storage = temp_storage;
+
+
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+
+      // Run
+      cudaErrchk(::cub::DeviceReduce::Reduce(d_temp_storage,
+                                             temp_storage_bytes,
+                                             x+ibegin,
+                                             sum_storage,
+                                             len,
+                                             ::cub::Sum(),
+                                             m_sum_init,
+                                             stream));
+
+      cudaErrchk(cudaStreamSynchronize(stream));
+      m_sum = *sum_storage;
+
+    }
+    stopTimer();
+
+    // Free temporary storage
+    deallocCudaDeviceData(temp_storage);
+    deallocCudaPinnedData(sum_storage);
+
+    REDUCE_SUM_DATA_TEARDOWN_CUDA;
+
+  } else {
+     getCout() << "\n  REDUCE_SUM : Unknown Cuda variant id = " << vid << std::endl;
+  }
+}
+
+template < size_t block_size >
+void REDUCE_SUM::runCudaVariantBlock(VariantID vid)
+{
+  const Index_type run_reps = getRunReps();
+  const Index_type ibegin = 0;
+  const Index_type iend = getActualProblemSize();
+
+  REDUCE_SUM_DATA_SETUP;
+
+  if ( vid == Base_CUDA ) {
+
+    REDUCE_SUM_DATA_SETUP_CUDA;
+
+    Real_ptr dsum;
+    allocCudaDeviceData(dsum, 1);
+
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+
+      initCudaDeviceData(dsum, &m_sum_init, 1);
+
+      const size_t grid_size = RAJA_DIVIDE_CEILING_INT(iend, block_size);
+      reduce_sum<block_size><<<grid_size, block_size,
+                  sizeof(Real_type)*block_size>>>( x,
+                                                   dsum, m_sum_init,
+                                                   iend );
+      cudaErrchk( cudaGetLastError() );
+
+      Real_type lsum;
+      Real_ptr plsum = &lsum;
+      getCudaDeviceData(plsum, dsum, 1);
+
+      m_sum = lsum;
+
+    }
+    stopTimer();
+
+    deallocCudaDeviceData(dsum);
+
+    REDUCE_SUM_DATA_TEARDOWN_CUDA;
+
+  } else if ( vid == RAJA_CUDA ) {
+
+    REDUCE_SUM_DATA_SETUP_CUDA;
+
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+
+      RAJA::ReduceSum<RAJA::cuda_reduce, Real_type> sum(m_sum_init);
+
+      RAJA::forall< RAJA::cuda_exec<block_size, true /*async*/> >(
+        RAJA::RangeSegment(ibegin, iend), [=] __device__ (Index_type i) {
+          REDUCE_SUM_BODY;
+      });
+
+      m_sum = sum.get();
+
+    }
+    stopTimer();
+
+    REDUCE_SUM_DATA_TEARDOWN_CUDA;
+
+  } else {
+     getCout() << "\n  REDUCE_SUM : Unknown Cuda variant id = " << vid << std::endl;
+  }
+}
+
+void REDUCE_SUM::runCudaVariant(VariantID vid, size_t tune_idx)
```

Comment:
Please insert some blank lines between code blocks to make it easier to parse visually